src/main/scala/com/fulcrumgenomics/bam/ComputeRnaSeqInsertSize.scala

```diff
@@ -0,0 +1,330 @@
+    |
+    |Also skips unpaired, failed QC, secondary, supplementary, pairs without both ends mapped, duplicates, and pairs whose
+    |reads map to different chromosomes.
+    |
+    |This tool requires each mapped pair to have the mate cigar (MC) tag and for read pairs not to overlap.  Use Picard's
+    |RevertOriginalBaseQualitiesAndAddMateCigar to add the former and Fgbio's ClipOverlappingReads to ensure the latter.
+  """,
+  group = ClpGroups.SamOrBam)
+class ComputeRnaSeqInsertSize
+(@arg(flag="i", doc="Input BAM file.") val input: PathToBam,
+ @arg(flag="r", doc="Input gene annotations in refFlat form.  Format described here: http://genome.ucsc.edu/goldenPath/gbdDescriptionsOld.html#RefFlat")
+ val refFlat: FilePath,
+ @arg(flag="o", doc="Output metrics file to which to write.") val output: FilePath
+) extends FgBioTool with LazyLogging {
+
+  Io.assertReadable(input)
+  Io.assertReadable(refFlat)
+  Io.assertCanWriteFile(output)
+
+  override def execute(): Unit = {
+    val progress = new ProgressLogger(logger, verb = "read", unit = 5e6.toInt)
+    val in = SamReaderFactory.make.open(input)
+    val refFlatSource = RefFlatSource(refFlat)
+    val geneOverlapDetector = new OverlapDetector[Gene](0, 0)
+    refFlatSource.foreach { gene => geneOverlapDetector.addLhs(gene, gene) }
+    val counters = PairOrientation.values().toList.map { pairOrientation => (pairOrientation, new NumericCounter[Long]()) }.toMap
+
+    in.foreach { rec =>
+      calculateInsertSize(rec=rec, overlapDetector=geneOverlapDetector) match {
+        case None             => Unit // ignore
+        case Some(insertSize) => counters(SamPairUtil.getPairOrientation(rec)).count(insertSize)
+      }
+      progress.record(rec)
+    }
+
+    refFlatSource.safelyClose()
+    in.safelyClose()
+
+    val metrics = PairOrientation.values().toList.map { pairOrientation =>
+      val counter = counters(pairOrientation)
+      val mean    = counter.mean()
+      InsertSizeMetric(
+        pairOrientation    = pairOrientation,
+        read_pairs         = counter.total,
+        mean               = mean,
+        median             = counter.median(),
+        min                = counter.map(_._2).min,
+        max                = counter.map(_._2).max,
+        standard_deviation = counter.stddev(m=mean)
+      )
+    }
+
+    Metric.write(metrics=metrics, path=output)
+  }
+
+  /** Calculates the insert size in transcript space. */
+  private def calculateInsertSize(rec: SAMRecord, overlapDetector: OverlapDetector[Gene]): Option[Int] = {
+    if (!rec.getReadPairedFlag || rec.getReadFailsVendorQualityCheckFlag || rec.isSecondaryOrSupplementary) return None
+    else if (rec.getReadUnmappedFlag || rec.getMateUnmappedFlag) return None
+    else if (rec.getDuplicateReadFlag || rec.getReferenceIndex != rec.getMateReferenceIndex) return None
+    else if (rec.getSecondOfPairFlag) return None
```

Comment:
A couple of thoughts:
1) You might want to just use a `FilteringSamIterator` to do a bunch of this for you and not have to have guards with returns here.
2) Do you want to also screen out reads with low mapq?